# Lab 7: Removing Duplicates & Handling Missing Values
## IBM Data Analytics Capstone Project 
### Objective: Remove duplicates, handle missing values, and normalize compensation data

In [1]:
import pandas as pd

# Load the dataset
file_path = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/n01PQ9pSmiRX6520flujwQ/survey-data.csv"
df = pd.read_csv(file_path)

print("Dataset loaded successfully!")
print("Shape:", df.shape)
df.head()

Dataset loaded successfully!
Shape: (65437, 114)


,ResponseId,MainBranch,Age,Employment,RemoteWork,Check,CodingActivities,EdLevel,LearnCode,LearnCodeOnline,...,JobSatPoints_6,JobSatPoints_7,JobSatPoints_8,JobSatPoints_9,JobSatPoints_10,JobSatPoints_11,SurveyLength,SurveyEase,ConvertedCompYearly,JobSat
0,1,I am a developer by profession,Under 18 years old,"Employed, full-time",Remote,Apples,Hobby,Primary/elementary school,Books / Physical media,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,2,I am a developer by profession,35-44 years old,"Employed, full-time",Remote,Apples,Hobby;Contribute to open-source projects;Other...,"Bachelor’s degree (B.A., B.S., B.Eng., etc.)",Books / Physical media;Colleague;On the job tr...,Technical documentation;Blogs;Books;Written Tu...,...,0.0,0.0,0.0,0.0,0.0,0.0,NaN,NaN,NaN,NaN
2,3,I am a developer by profession,45-54 years old,"Employed, full-time",Remote,Apples,Hobby;Contribute to open-source projects;Other...,"Master’s degree (M.A., M.S., M.Eng., MBA, etc.)",Books / Physical media;Colleague;On the job tr...,Technical documentation;Blogs;Books;Written Tu...,...,NaN,NaN,NaN,NaN,NaN,NaN,Appropriate in length,Easy,NaN,NaN
3,4,I am learning to code,18-24 years old,"Student, full-time",NaN,Apples,NaN,Some college/university study without earning ...,"Other online resources (e.g., videos, blogs, f...",Stack Overflow;How-to videos;Interactive tutorial,...,NaN,NaN,NaN,NaN,NaN,NaN,Too long,Easy,NaN,NaN
4,5,I am a developer by profession,18-24 years old,"Student, full-time",NaN,Apples,NaN,"Secondary school (e.g. American high school, G...","Other online resources (e.g., videos, blogs, f...",Technical documentation;Blogs;Written Tutorial...,...,NaN,NaN,NaN,NaN,NaN,NaN,Too short,Easy,NaN,NaN


In [2]:
# Count duplicate rows
duplicate_count = df.duplicated().sum()
print("Number of duplicate rows:", duplicate_count)

# Display first few duplicates if any
if duplicate_count > 0:
    print("\nFirst few duplicate rows:")
    print(df[df.duplicated()].head())
else:
    print(" No duplicate rows found in this dataset!")

Number of duplicate rows: 0
 No duplicate rows found in this dataset!


In [3]:
# Remove duplicates (verification step)
df_clean = df.drop_duplicates()

print("Original shape:", df.shape)
print("After drop_duplicates shape:", df_clean.shape)
print("Rows removed:", df.shape[0] - df_clean.shape[0])
print(" Dataset is already free of duplicates!")

Original shape: (65437, 114)
After drop_duplicates shape: (65437, 114)
Rows removed: 0
 Dataset is already free of duplicates!


In [4]:
# Identify missing values for all columns
missing_values = df.isnull().sum()
missing_percent = (df.isnull().sum() / len(df)) * 100

missing_df = pd.DataFrame({
    'Missing Values': missing_values,
    'Percentage': missing_percent
})

# Show only columns with missing values
missing_df = missing_df[missing_df['Missing Values'] > 0].sort_values('Missing Values', ascending=False)

print("Columns with missing values:")
print(missing_df.head(20))

Columns with missing values:
                               Missing Values  Percentage
AINextMuch less integrated              64289   98.245641
AINextLess integrated                   63082   96.401119
AINextNo change                         52939   80.900714
AINextMuch more integrated              51999   79.464217
EmbeddedAdmired                         48704   74.428840
EmbeddedWantToWorkWith                  47837   73.103901
EmbeddedHaveWorkedWith                  43223   66.052845
ConvertedCompYearly                     42002   64.186928
AIToolNot interested in Using           41023   62.690832
AINextMore integrated                   41009   62.669438
Knowledge_9                             37802   57.768541
Frequency_3                             37727   57.653927
Knowledge_8                             37679   57.580574
ProfessionalTech                        37673   57.571405
Knowledge_7                             37659   57.550010
Knowledge_6                             375

In [5]:
# Check missing values in EdLevel specifically
print("Missing values in EdLevel:", df['EdLevel'].isnull().sum())
print("\nEdLevel value counts:")
print(df['EdLevel'].value_counts())

# Impute with most frequent value
most_frequent_edlevel = df['EdLevel'].value_counts().idxmax()
print("\nMost frequent EdLevel:", most_frequent_edlevel)

df['EdLevel'].fillna(most_frequent_edlevel, inplace=True)
print("Missing values in EdLevel after imputation:", df['EdLevel'].isnull().sum())
print(" EdLevel missing values imputed successfully!")

Missing values in EdLevel: 4653

EdLevel value counts:
EdLevel
Bachelor’s degree (B.A., B.S., B.Eng., etc.)                                          24942
Master’s degree (M.A., M.S., M.Eng., MBA, etc.)                                       15557
Some college/university study without earning a degree                                 7651
Secondary school (e.g. American high school, German Realschule or Gymnasium, etc.)     5793
Professional degree (JD, MD, Ph.D, Ed.D, etc.)                                         2970
Associate degree (A.A., A.S., etc.)                                                    1793
Primary/elementary school                                                              1146
Something else                                                                          932
Name: count, dtype: int64

Most frequent EdLevel: Bachelor’s degree (B.A., B.S., B.Eng., etc.)
Missing values in EdLevel after imputation: 0
 EdLevel missing values imputed successfully!


/var/folders/l6/_t_zyy_55010xg3998z5sfyc0000gn/T/ipykernel_82342/801968193.py:10: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['EdLevel'].fillna(most_frequent_edlevel, inplace=True)


In [6]:
# Check missing values in ConvertedCompYearly
print("Missing values in ConvertedCompYearly:", df['ConvertedCompYearly'].isnull().sum())
print("Total rows:", len(df))

# Handle missing values - fill with median (better than mean for salary data)
median_comp = df['ConvertedCompYearly'].median()
print("\nMedian CompYearly:", median_comp)

df['ConvertedCompYearly'] = df['ConvertedCompYearly'].fillna(median_comp)
print("Missing values after imputation:", df['ConvertedCompYearly'].isnull().sum())

# Summary statistics
print("\nConvertedCompYearly Summary:")
print(df['ConvertedCompYearly'].describe())
print(" Compensation data normalized successfully!")

Missing values in ConvertedCompYearly: 42002
Total rows: 65437

Median CompYearly: 65000.0
Missing values after imputation: 0

ConvertedCompYearly Summary:
count    6.543700e+04
mean     7.257636e+04
std      1.122207e+05
min      1.000000e+00
25%      6.500000e+04
50%      6.500000e+04
75%      6.500000e+04
max      1.625660e+07
Name: ConvertedCompYearly, dtype: float64
 Compensation data normalized successfully!


## Summary - Lab 7: Data Wrangling

### Key Findings:
- **Duplicates:** 0 duplicate rows found in the clean dataset
- **Missing Values:** Multiple columns had significant missing data
  - AINext columns had up to 98.2% missing values
  - ConvertedCompYearly had 64.2% missing (42,002 rows)
  - EdLevel had 4,653 missing values (7.1%)

### Actions Taken:
- **EdLevel:** Imputed 4,653 missing values with most frequent value 
  → "Bachelor's degree"
- **ConvertedCompYearly:** Imputed 42,002 missing values with median 
  → $65,000

### Compensation Insights:
- Median annual compensation: $65,000
- Mean annual compensation: $72,576
- Max compensation: $16,255,600 (likely outlier)